### DISCUSSION TITLE
problem with @state_trigger (help! :))

### DISCUSSION THREAD

**Discussion by wsw70** [custom-components/pyscript#292] [Q&A]
I created the following helper function (in `modules`), it does an operation once the state of a sensor changes:

```
def rf433(value=None, operation=None, entity_id=None):
    if not all([value, operation, entity_id]):
        log.error("payload, operation, entity_id parameter missing")
    log.info(f"setting up RF433 device {value} to trigger {entity_id}")

    @state_trigger("sensor.rf433")
    def message(**data):
        log.info(data)
        if data['value'][-6:] == value:
            operation(entity_id=entity_id)
            log.info(f"{entity_id} triggered by switch {value}")

    return message
```

it is actually set-up in a script:

```
triggers = [
    volets.volets_time(entity_id="cover.parents_volets", what="up", when="09:00"),
    volets.volets_time(entity_id="cover.parents_volets", what="down", when="sunset+1h"),
    switch.rf433(value="1CA0B8", entity_id="switch.salon_lampe_haute", operation=homeassistant.toggle)
]
```

Upon pressing a RF433 button, a Sonoff RF Bridge sends a MQTT message and the following sensor is set:

```
  - platform: mqtt
    state_topic: "rfbridge-1/rfin"
    name: rf433
    unique_id: rf433
```

I also have a catch-all logging function for any RF433 message:

```
# log activity from RF433 devices
log.info("starting logging of RF433 devices")
@state_trigger("sensor.rf433")
def rf433_log_message(**data):
    # sometimes we get a phantom message
    if data['value'][-6:] == '000000':
        pass
    log.info(f"RF433 device {data['value'][-6:]} was pressed")
```

**The mystery: `switch.salon_lampe_haute` is toggled but the logs are not displayed** I just see the message from 
- the initialization
- the catch-all function:

```
2022-01-01 20:59:17 INFO (MainThread) [custom_components.pyscript.scripts.automation.parents] setting up RF433 device 1CA0B8 to trigger switch.salon_lampe_haute
2022-01-01 20:59:43 INFO (MainThread) [custom_components.pyscript.scripts.automation._startup_.rf433_log_message] RF433 device 1CA0B8 was pressed
```

Of course my first thought was that this action is defined elsewhere, so I changed `switch.salon_lampe_haute` to another lamp, restarted HA and pressing the button now toggles that other lamp.

**My question: can I have several `@state_trigger()` on the same entity?** I guess I can because I see a log from one of the functions, and the action form another. Is it possible that there is some conflict in th elogging functions? (the first one wins, and the second one is discarded?)

**Comment by beiarn** [comment]
You can have several triggers. Maybe it won't be very efficient. Anyway, there is limitations to actions.
Regarding your logger problem. have you considered EVENT_STATE_CHANGED?
look up the eventbased triggers on the wiki.

Under is an example on how i use "EVENT_STATE_CHANGED" in my energycost app:
adapt it to your need, you dont need to use @state_active.

@event_trigger(EVENT_STATE_CHANGED, "entity_id.startswith('pyscript.consumption_highest_thismonth')") #Henter inn energimålerverdier 
@state_active("input_boolean.nettleige_beregninger_aktiv == 'on'")
def monitor_state_highest_thismonth(entity_id=None, new_state=None, old_state=None):
    global highest_thismonth_energymeter

    old_value = old_state.state if old_state else None
    #log.info(f"entity {entity_id} changed from {old_value} to {new_state.state}")
    highest_thismonth_energymeter = float(new_state.state) # kWh format
    #log.info(f" Målerstanden er nå: {highest_thismonth_energymeter} kWh")

**Comment by dlashua** [comment]
You can have MANY functions with the same `@state_trigger`.

I think, in this case, the reason you're not seeing the logging is because the value of the `sensor.rf433` is not changing. Try this instead...

```python
@state_trigger(
    'True',
    watch = ['sensor.rf433']
)
def rf433_log_message(**data):
  # .....
```

As a side note, I have found odd behavior (see #281) regarding modules that return trigger functions. I don't think that's interfering in this case, but, there may be other side effects I've not discovered yet.

**Reply by beiarn** [reply]
Nice. Learned something new. :)
